### PR TITLE
Implements `json_metadata` RPC call

### PR DESCRIPTION
**Pull request by bkchr**

### CODE REVIEW

File: demo/runtime/src/lib.rs [hunk 1]
```diff
@@ -229,9 +229,20 @@ pub type CheckedExtrinsic = generic::CheckedExtrinsic<AccountId, Index, Call>;
 pub type Executive = executive::Executive<Runtime, Block, Balances, Balances,
 	(((((), Council), Democracy), Staking), Session)>;
 
+impl_json_metadata!(for Runtime with modules
```
Comment (rphmeier) [inline demo/runtime/src/lib.rs]: style: "all-or-nothing" indentation with close-paren on its own line.

File: substrate/runtime-support/src/metadata.rs [hunk 2]
```diff
@@ -0,0 +1,186 @@
+/// Implements the json metadata support for the given runtime and all its modules.
```
Comment (rphmeier) [inline substrate/runtime-support/src/metadata.rs]: needs license header

File: substrate/client/src/client.rs [hunk 3]
```diff
@@ -251,6 +251,13 @@ impl<B, E, Block> Client<B, E, Block> where
 		&self.executor
 	}
 
+	/// Returns the runtime metadata as JSON.
+	pub fn json_metadata(&self, id: &BlockId<Block>) -> error::Result<String> {
+		self.executor.call(id, "json_metadata",&[])
+			.and_then(|r| String::decode(&mut &r.return_data[..])
+					  .ok_or("Metadata decoding failed".into()))
```
Comment (gavofyork) [inline substrate/client/src/client.rs]: check style guide on indentation. basically, it's only ever a single tab at a time, never *ever* mix tabs + spaces.

File: substrate/runtime-support/src/metadata.rs [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+	pub trait Trait {
+		 type PublicAux;
+	}
+
+	impl Trait for TestRuntime {
+		type PublicAux = u32;
+	}
+
+	impl_json_metadata!(for TestRuntime with modules
+						event_module::Module,
```
Comment (gavofyork) [inline substrate/runtime-support/src/metadata.rs]: more bad intentation

**Comment by bkchr** [comment]
@gavofyork The wasm stuff is compiled with a nightly compiler and the native stuff is compiled with a stable compiler? Just asking, because I thought of using `const fn`, but that is still unstable.

**Comment by gavofyork** [comment]
that's right - wasm with nightly (it doesn't compile using stable yet)

File: substrate/client/src/client.rs [hunk 5]
```diff
@@ -711,4 +734,29 @@ mod tests {
...
+		assert!(builder.push_transfer(Transfer {
+			from: Keyring::Eve.to_raw_public().into(),
+			to: Keyring::Alice.to_raw_public().into(),
+			amount: 42,
+			nonce: 0,
+		}).is_err());
+
+		client.justify_and_import(BlockOrigin::Own, builder.bake().unwrap()).unwrap();
+
+		assert_eq!(client.json_metadata(&BlockId::Number(1)).unwrap(), r#"{ "events": "events" }"#);
```
Comment (gavofyork) [inline substrate/client/src/client.rs]: what does this test?

File: substrate/runtime-support/src/metadata.rs [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+/// In this example, just `MODULE3` implements the `Storage` trait.
+#[macro_export]
+macro_rules! impl_json_metadata {
+	(
+		for $runtime:ident with modules
+		$( $rest:tt )*
+	) => {
+		impl $runtime {
+			pub fn json_metadata() -> $crate::metadata::Vec<$crate::metadata::JSONMetadata> {
+					__impl_json_metadata!($runtime;
```
Comment (gavofyork) [inline substrate/runtime-support/src/metadata.rs]: sueprfluous tab

File: substrate/client/src/client.rs (hunk 5 above)
Comment (bkchr) [reply substrate/client/src/client.rs]: That just tests that the `json_metadata` function of the `test-client` is getting called.

**Pull request merged**